### DISCUSSION TITLE
Getting started With NomadNetwork My perspective

### DISCUSSION THREAD

**Discussion by ve3jec** [markqvist/Reticulum#224] [Nomad Network]
Hello all from Ontario, Canada.

So after reading Dualfuel671's [Getting started with NomadNetwork](https://github.com/markqvist/Reticulum/discussions/166) i decided it was helpful to my adventure and i should likely do one myself as i am clearly less knowledgeable about all things CLI and it may help other that are "newbies" to Linux. 

So lets list the hardware on hand i have to tinker with
1- Rpi 3b+ bare
1- Rpi 3b+ W/7" touchscreen, with a buck converter and cap added to allow it to be happy with 7-32VDC input. Added as the screen is a power hog. 
1- Rpi Zero W
2- TNC-X (simple kiss TNC identical to the MFJ-1270X 
1- TNCPi
Various VHF and UHF radios to use with the TNC's 

My first steps were to try getting Nomadnet installed into Linux Mint Xfce 21.1 64bit.
That went painlessly, so i added a second VM and now can talk back and forth using the testnet.

First hurdle was figuring out that 
`cd /.nomadnetwork/storage` 
needs the / on it so its 
`cd /.nomadnetwork/storage/`
once i figured that out i was good to go. 

Figured out how to save nodes, and set peers as trusted
**Figured out it REALLY needs a copy/paste function for those addresses**
Noticed it could use having all the shortcuts dumped into a config file, so someone with VM issues or similar could pull up the file and change them. 

Guide could use some additions, like a link or cart for the colours. EVERY thing google found me did not line up on the colour choice, so i ended up just doing the edit, save, test on repeat to figure out a couple.  Same go's with Micron, my google fu sucked on this all i could get back was micron device hardware. Likely me not using the right search term.

Micron Vs html, while similar in structure, i do not find them alike, mind you its been 2 decades <img src="lol.gif">. did a basic page on one VM (Node_Test_2) and hit a wall with file downloads. cant make it work, but i know it works as i pulled an image from the "Unsigned Node".  But i will come back around to this in a new post in this topic.

My next steps are getting the 3b+ pi's setup and updated, install Nomadnet and try getting them to speak to the TNC's  for the short term i am just going to wire the tnc's to each other and skip the RF setup since my radios are all 5W+ and i only have one Dummy load. 

I do enjoy catching new Nodes and reading their pages, seeing what they have came up with. Was thinking last night it would be neat to sort a way to pull text news articles into a page.  

You will find i jump around some, and its mostly aimed at keeping my frustration low and interest high. 
Cheers,
Joe

**Comment by ve3jec** [comment]
Setting up the Raspi 3b+ headless
Using Raspi OS lite 64bit
I cheat and use a monitor for setup instead of editing files on the boot before i use it. 
Once its all setup i move over to putty from there on. 

`sudo apt update`
`sudo apt upgrade -y`

Pips not installed be default so i added it
`sudo apt install python3-pip`
got some path not added warnings on its complete.
tried `nomadnet` and no go
hmm lets reboot
`sudo reboot`

Once its back
`nomadnet` and it dose its first start up routine
loads into the guide like it should.

head to the **_Network_** tab and  change my Name from anonymous peer and save it.
head to the **_Config_** tab and make note of the config path 
/home/raspi/.nomadnetwork/config
that means Reticulum's path will be
/home/raspi/.reticulum/config

and i quit NomadNet

Lets move to the Reticulum folder and have a loo
`cd /home/raspi/.reticulum/`

fair enough that work so i back out
`cd`

lets edit the Reticulum config file to get onto the Testnet
`nano /home/raspi/.reticulum/config`

dropped to the bottom and added the 

```
# TCP/IP interface to the Dublin Hub
  [[RNS Testnet Dublin]]
    type = TCPClientInterface
    enabled = yes
    target_host = dublin.connect.reticulum.network
    target_port = 4965

# TCP/IP interface to the Frankfurt Hub
  [[RNS Testnet Frankfurt]]
    type = TCPClientInterface
    enabled = yes
    target_host = frankfurt.connect.reticulum.network
    target_port = 5377

# Interface to I2P Hub A
  [[RNS Testnet I2P Hub A]]
    type = I2PInterface
    enabled = yes
    peers = pmlm3l5rpympihoy2o5ago43kluei2jjjzsalcuiuylbve3mwi2a.b32.i2p

# Interface to I2P Hub B
  [[RNS Testnet I2P Hub B]]
    type = I2PInterface
    enabled = yes
    peers = iwoqtz22dsr73aemwpw7guocplsjjoamyl7sogj33qtcd6ds4mza.b32.i2p
```
Save that and exit i generally do it in 2 steps CTRL-O hit enter, then CTRL-X

reload NomadNet
`nomadnet`

So i wasn't sure with my VM's with the I2P errors and just commented it out in the Reticulum config, and i get same errors now. 
Most likely the required packages are not installed.  so lets comment it out for now and i will sort getting it installed another time.

exit NomadNet and open the editor for the reticulum config
you can retype it all or just arrow up twice and its there ready

add the # to comment out all the I2P hub lines 
```
# Interface to I2P Hub A
#  [[RNS Testnet I2P Hub A]]
#    type = I2PInterface
#    enabled = yes
#    peers = pmlm3l5rpympihoy2o5ago43kluei2jjjzsalcuiuylbve3mwi2a.b32.i2p

# Interface to I2P Hub B
#  [[RNS Testnet I2P Hub B]]
#    type = I2PInterface
#    enabled = yes
#    peers = iwoqtz22dsr73aemwpw7guocplsjjoamyl7sogj33qtcd6ds4mza.b32.i2p
```
Save it and reload nomadnet
Now the only warning i get is the could not auto select a propagation node
Network tab and add a url with C-u (nomadnet is showing short form for CTRL-U)  and i entered the Dublin hub address from the [NomadNet github](https://github.com/markqvist/NomadNet) info
Page loads like expected.

So i will leave this to run and wait for some nodes to show up in the Announce Stream, and save the RNS Dubblin and RNS Frankfurt nodes 
That can be done by clicking on the address to view the info and hit save
then i hit C-L and it shows saved nodes, now i select the node i just saved and hit info and can select use as default propagation node and save that.

That should take care of the log errors for now. 
Don't forget to C-L back to the announce stream to see new peers and node show up. 


So some thing i have noted to this point.
Need a C-?? something to save a node you are viewing so you don't need to wait for it to announce itself

Putty doesn't show the P and N icons on its default fonts, nor spacemono nerd font. <<- - works once you enable nerd in the config for nomad

This is long enough now lol
Cheers

**Comment by ve3jec** [comment]
So lets move to Test_Node_2

I have been trying to get a file download to work just a simple test.txt file and was a fail.

I have come to the conclusion that the directory needs to be named **files** but had to be called upon as **file**  

once that worked i moved to trying to load a page, named page1.mu 
This works on the host node only, loads the page fine, but is not accessible from other peers. 
```
`cWelcome to my `!`FfffNode`f`!


Just working my way into learning how to get
around and make pages work as they should.


`F2f0`_`[link `ebafbf6ef7ba1f58d9f525c056abe766:/file/test.txt]`_`f to test.txt

`F2f0`_`[Page `ebafbf6ef7ba1f58d9f525c056abe766:/page/page1.mu]`_`f Page 1
```
Spacing out the code changed nothing aside form its look with the underlined spaces.
tried with page, pages, files, file. ended up with a saved copy of the page on the node, but the peer would touch that link, just request failed. 

yup

Time to go make supper and watch a movie i think.
Cheers
Joe

**Comment by Swissbandit** [comment]
Hi,nice to see intrest in rns growing

Do restart your nomadnet when you finish changing dl files and add more pages,as it will become only visible after reloading your node,
while to see for yourself can just refresh the page and its visible for you,but not others.
you are correct in that files go to the files directory and are linked with file/... its important that if you share a .txt file its also the same in the link path

you can share any file,but need to crate a link to the right file in files folder matching its name and extention.

also if a file or page is on the same node it does not need the nodeadress in the links

a simple [link to a page ` :/page/index.mu] `
is fine on the same node
if you link to another nodes page youll simply attach the adress ahead of the link path

same with files

[DL my picture here `:/file/pic.jpg]`

also dont forget the pages get chached,if you want to see the new version of an already chached page just klick into it and use ctrl + r to refresh the page(it will be reloaded using a new connection to the node hosting the page)

my Swisslibertarian Node has some more infos and examples on those things, 
also in the guide tab of nomadnet is some more info on how to create micron pages.

I hope this will help you further in constructing pages and content for your Nodes

feeel free to ask here or on the ret

**Reply by markqvist** [reply]
Thanks for sharing these useful tips and tricks @Swissbandit!

It's useful also to know that you can control the caching behaviour of pages by including the cache header in the page. To do so, the first line of the page must be like this `#!c=X`, where X is the number of seconds the client should cache the page for.

**Comment by ve3jec** [comment]
I will hop into setting and kill the announce on the node and do my testing rebooting the node each integration, no point in hammering the system with an announce every couple of minutes or less.

Ok so that seems to have solved the page loading issue, and linking issue i had (with not using the node address) but i will delve into it later when the suns go's down. 

I never noticed in the guide if it is there, but can the node reboot on page edit be added in a distinctive colour so others see it when they get started?


Anyways Much appreciation to you both for chiming and pointing me down the right path.
Cheers
Joe

**Reply by markqvist** [reply]
You only need to reboot the program when adding *new* pages. It scans the available ones on program start, so you can *edit* without restarting the program, but if you add new ones, you will have to.

> I never noticed in the guide if it is there, but can the node reboot on page edit be added in a distinctive colour so others see it when they get started?

Good idea. I'll make it more clear and visible in the guide!

**Reply by ve3jec** [reply]
Ok thanks for the clarification!

Cheers

**Comment by ve3jec** [comment]
@markqvist  Is their a place to access the guide outside the program. i'd like to be able to reference it with either a notepad or pdf on whatever device i chose exclusive of nomadnet if needed. 

I know there is a PDF for Reticulum, but looking and searching has turned up nothing for Nomadnet. 
Cheers
Joe

PS.  Raspi/tnc  wired to raspi/tnc works, haven't wrote it up yet tough.

**Reply by Swissbandit** [reply]
here is the source file https://github.com/markqvist/NomadNet/blob/master/nomadnet/ui/textui/Guide.py

**Reply by markqvist** [reply]
I'll see if I can get it all compiled into a PDF and HTML version for easier referencing. If someone beats me to it, I will be very happy though ;)

**Reply by ve3jec** [reply]
Thanks for the link, i looked in the storage folders of the install, but never crossed my mind in the github for it.

**Reply by Swissbandit** [reply]
[nomadguide.txt](https://github.com/markqvist/Reticulum/files/10775061/nomadguide.txt)
i wanted to update my translations so i started with the guide file first
(also to learn about the new fields) while the part with making pages and letting only allowed id `s see a page has not worked here so far.

**Comment by ve3jec** [comment]
So this install will be about getting the TNC's working. Its pretty straight forward

In my testing setup i will be omitting the RF out of this setup and test. Mainly for simplicity of testing as RF requires much more hardware setup.

Wiring is simple 
Both Raspi's are connected to Wi-Fi 
Headless raspi is wired to a buck converted via USB input with the converter outputting 5.2V 
Raspi with the 7" touchscreen is wired to a buck converter, that feeds a 200v 560uF cap to smooth the power, then hard wired to the Raspi and screen power inputs. (i found the cap was needed to keep it happy over the long term)
both TNC's are plugged in via USB to the raspi

Wiring between TNC's is simply a ground between them and tx to rx cross over in each direction. 

I opened each pi's reticulum config and added the KISS interface code  editing the `speed` line from 115200, to 9600 as i have the jumpers set in the TNC. 

```
[[Packet Radio KISS Interface]]
  type = KISSInterface
  interface_enabled = True

  # Serial port for the device
  port = /dev/ttyUSB1

  # Set the serial baud-rate and other
  # configuration parameters.
  speed = 9600
  databits = 8
  parity = none
  stopbits = 1

  # Set the modem preamble.
  preamble = 150

  # Set the modem TX tail.
  txtail = 10

  # Configure CDMA parameters. These
  # settings are reasonable defaults.
  persistence = 200
  slottime = 20

  # You can configure the interface to send
  # out identification on the channel with
  # a set interval by configuring the
  # following two parameters. The KISS
  # interface will only ID if the set
  # interval has elapsed since it's last
  # actual transmission. The interval is
  # configured in seconds.
  # This option is commented out and not
  # used by default.
  # id_callsign = MYCALL-0
  # id_interval = 600

  # Whether to use KISS flow-control.
  # This is useful for modems that have
  # a small internal packet buffer, but
  # support packet flow control instead.
  flow_control = false
```
 
For testing i attached a display to the headless setup so i could drop the Wi-Fi on both and still use them.
Launched nomadnet and watched the LED's on the TNC dance. 
Could send announcements back and forth no issue.
At that point i shutdown as i was out of time and had to go.

The next test will be setting one as a node with a few pages to navigate, some with a decent chunk of text and see how i fair for transfer time between the 2. 

Cheers
Joe

**Reply by markqvist** [reply]
Looking forward to hearing your further explorations in this!

**Comment by ve3jec** [comment]
So transfer from the node to the peer via TNC as setup above
Index.mu 95B total, 95B in 4.69 seconds
0.01 Seconds over Wi-Fi
I have put roughly 2 chapters of Killing Floor by Lee Child on a page to load 
39.9KB  in 114.02 Seconds or 955b/s  
0.06 Seconds over Wi-Fi to load that page 


The hole book is 797KB or 817,050 bytes so at that rate it would take 855.55 Seconds or a little over 14 minutes to transfer barring any packet loss or other interruptions. 

The whole Reacher library is 51.3mb of epub files and they seem to be less space then a .txt file by almost half.
so epub to TXT and then made into a library it would take 1686 minutes (28 hours 6 Minutes) to transfer that whole set over the TNC lol
or 940 Minutes for the epub files as they are, or 889 Minutes if the epub files are compressed into a .zip file at the best it can do.
 


I'd be interested in seeing Pactor-4 rates tested, should top out around 5500baud but the hardware is pretty costly at $1700 each here ($1921 after tax) 

i wonder how well dedicated uplink/downlink would work for the TNC's (read Rnode's will do this), i wouldn't expect it to be twice as fast maybe 25% faster overall. One could make it work with 2 Radio's and 2 SDR's or 4 radios. would require both VHF and UHF radios though, or repeater cans so you could be on a single band. At that price point and complexity i would expect other methods to be better.

**Reply by markqvist** [reply]
Interesting stuff, thanks for sharing the results here!

> 39.9KB in 114.02 Seconds or 955b/s

That's a pretty darned good effective rate (which takes into account the overhead of everything) over 1200-baud AFSK!

> or 940 Minutes for the epub files as they are, or 889 Minutes if the epub files are compressed into a .zip file at the best it can do.

For files/chunks up to around 16 megabytes, it will actually probably be better to send them directly as text, as Reticulum automatically compresses resource transfers with the highest possible level of bzip compression if it can save bandwidth. For very large files, it *might* be better to compress externally first, but from what I have seen, there is not really much difference, unless there are very certain types of compressible structures in the data that extend across the 16 megabyte chunk boundaries.

> I'd be interested in seeing Pactor-4 rates tested, should top out around 5500baud but the hardware is pretty costly at $1700 each here ($1921 after tax)

I really hope to someday have an open source solution to this $2000 dollar nonsense for a proper narrow-band modem. An ESP32 and a couple of switchable filters ($2 in passives) could probably do it.

**Reply by ve3jec** [reply]
I intend on running the test again using the AX.25 mode, purely to see the difference the extra overhead makes.
I also intend on running the test RF, and can do both KISS, and KISS AX.25 in that arrangement. 

I Run my link speed (PC to TNC) at 9600, and i am curious to see if changed to 1200 what issues that causes if any. 
RF tests will most likely be done at 25W (lowest setting) on Kenwood TM-271 and TM-281 radios.
